### ISSUE TITLE
java.util.Optional is considered mutable

### ISSUE THREAD

**Issue by seanjreilly** [MutabilityDetector/MutabilityDetector#57]
After looking at the test results once issue #56, I noticed this in one of the error messages from the test:

`Field can have a mutable type (java.util.Optional) assigned to it. [Field: cause, Class: com.equalexperts.logging.LogicalLogRecord]`

I think that instances of java.util.Optional should probably be considered immutable so long as the class they are referencing is immutable. It's not in my case, as java.lang.Throwable isn't really immutable, but I suspect that's not why the error was raised.

Thoughts?

**Comment by Grundlefleck** [comment]
It's been almost a year... but I think I've finally got around to building the mechanism to support this. All that's left is picking the testing API. 
Outcome:
 - a `java.util.Optional` field is immutable if the contained element is immutable. Otherwise if it's a concrete type e.g. `Optional<Date>` or a generic type e.g. `Optional<A>`, it's mutable.
 - if you want to test that the class is immutable as long as the generic type variable is immutable, this can be allowed with `provided("MY_GENERIC_TYPE").isAlsoImmutable()`

I think it would be better not to use `provided` in this case, but will leave it for now and mark this issue closed.

@seanjreilly sorry it's been so long, hopefully you're still using Mutability Detector :)

**Comment by seanjreilly** [comment]
Hi. We are still using mutability detector, and we're very happy with it! We've coded around the limitations with Optional, but we'd love to be able to remove that. One question: I can't find a release newer than 0.9.2, which dates from before this issue... is there a newer release we should try?

**Comment by Grundlefleck** [comment]
@seanjreilly There's nothing newer than 0.9.2 at the moment. I plan to release 0.9.3, with the Optional fix, as soon as I finish #39. It would be really great if you could test a snapshot before publishing 0.9.3, is that something you can do?

**Comment by seanjreilly** [comment]
Hi,

No problem. I'm happy to run our tests with a snapshot. I will hopefully have time for that this weekend.

**Comment by seanjreilly** [comment]
I think I found a slight issue with the Optional fix. It seems to work well, but it doesn't seem to pass without adding `AllowedReason.provided(Optional.class).isAlsoImmutable()` to the test definition.

I'm not sure if this is a bug or not, but I've created a simple git repo illustrating how I got the MutabilityDetector test to work with the current snapshot, and how I think it ought to work.

The repo is [here] (https://github.com/seanjreilly/mutability-detector-issue-57).

Am I totally off base with this thought? Also, is this the right issue, or should I open a new one?

**Comment by Grundlefleck** [comment]
Hi @seanjreilly, many thanks for testing, you found a bug. Thankfullyit's just an error in the default configuration. You will get the behaviour you want with the following code:

```
    @Test
    public void how_it_would_work_if_wasnt_an_idiot_and_misconfigured_optional_in_the_default_configuration() throws Exception {
        MutabilityAsserter asserter = MutabilityAsserter.configured(new ConfigurationBuilder() {
            @Override
            public void configure() {
                hardcodeAsImmutableContainerType("java.util.Optional");
                mergeHardcodedResultsFrom(Configurations.JDK_CONFIGURATION);
            }
        });

        asserter.assertInstancesOf(ImmutableClass.class, areImmutable(), AllowedReason.provided(ImmutableClass.Bar.class).areAlsoImmutable());
    }
```
This test passes with the snapshot, but I'll be sure to fix the default config before releasing 0.9.3.

Good catch!

**Comment by Grundlefleck** [comment]
Just released 0.9.3 that should behave the way you expect with `Optional`.

**Comment by seanjreilly** [comment]
Just checked out the new version, and it works swimmingly. I actually use the mutability detector in a few projects — I've updated one, and will do the other very soon.

Thanks!
